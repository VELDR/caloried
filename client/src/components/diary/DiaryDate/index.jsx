import PropTypes from 'prop-types';
import { injectIntl } from 'react-intl';
import { useDispatch } from 'react-redux';
import { ChevronLeft, ChevronRight } from '@mui/icons-material';
import { Popover, Tooltip } from '@mui/material';
import { useState } from 'react';
import Calendar from 'react-calendar';
import 'react-calendar/dist/Calendar.css';
import { setDate } from '@pages/Diary/actions';

import classes from './style.module.scss';

const DiaryDate = ({ currentDate, setCurrentDate, intl: { formatMessage } }) => {
  const dispatch = useDispatch();
  const [anchor, setAnchor] = useState(null);

  const handlePrevDate = () => {
    setCurrentDate((prevDate) => {
      const newDate = new Date(prevDate.setDate(prevDate.getDate() - 1));
      dispatch(setDate(newDate));
      return newDate;
    });
  };

  const handleNextDate = () => {
    setCurrentDate((prevDate) => {
      const newDate = new Date(prevDate.setDate(prevDate.getDate() + 1));
      dispatch(setDate(newDate));
      return newDate;
    });
  };

  const handlePopoverOpen = (event) => {
    setAnchor(event.currentTarget);
  };

  const handlePopoverClose = () => {
    setAnchor(null);
  };

  const formattedDate = currentDate.toLocaleDateString('en-US', {
    weekday: 'long',
    day: 'numeric',
    month: 'long',
    year: 'numeric',
  });

  return (
    <div className={classes.container} data-testid="diary-date">
      <ChevronLeft onClick={handlePrevDate} />

      <Tooltip title={formatMessage({ id: 'app_select_a_date' })}>
        <div className={classes.date} onClick={handlePopoverOpen} data-testid="calendar-date">
          {formattedDate}
        </div>
      </Tooltip>
      <ChevronRight onClick={handleNextDate} />

      <Popover
        open={Boolean(anchor)}
        anchorEl={anchor}
        onClose={handlePopoverClose}
        anchorOrigin={{
          vertical: 'bottom',
          horizontal: 'center',
        }}
        transformOrigin={{
          vertical: 'top',
          horizontal: 'center',
        }}
        data-testid="calendar-popover"
      >
        <Calendar
          value={currentDate}
          onChange={(newDate) => {
            setCurrentDate(newDate);
            dispatch(setDate(newDate));
          }}
        />
      </Popover>
    </div>
  );
};

DiaryDate.propTypes = {
  currentDate: PropTypes.instanceOf(Date),
  setCurrentDate: PropTypes.func,
  intl: PropTypes.object,
};

export default injectIntl(DiaryDate);
