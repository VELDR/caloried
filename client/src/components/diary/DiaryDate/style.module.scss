@import '@styles/base';

.container {
  @include flex-items-center;
  svg {
    border-radius: 50%;
    cursor: pointer;
    padding: 0.25rem;
    &:hover {
      color: var(--color-primary);
      @include translucent-background($color-primary, 0.2);
    }
  }
  .date {
    cursor: pointer;
    text-align: center;
    width: 16rem;
    white-space: nowrap;
    &:hover {
      color: var(--color-text-secondary);
    }
  }
}
