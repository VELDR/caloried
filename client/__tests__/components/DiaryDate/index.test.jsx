import DiaryDate from '@components/diary/DiaryDate';

import { fireEvent, render } from '@utils/testHelper';

let wrapper;

const mockProps = {
  currentDate: new Date('2023-01-01'),
  setCurrentDate: jest.fn(),
  intl: {
    formatMessage: jest.fn(),
  },
};

beforeEach(() => {
  wrapper = render(<DiaryDate {...mockProps} />);
});

describe('DiaryDate Component', () => {
  it('should render component correctly', () => {
    const { getByTestId } = wrapper;
    expect(getByTestId('diary-date')).toBeInTheDocument();
  });

  it('should open popover on date click', () => {
    const { getByTestId } = wrapper;
    const dateElement = getByTestId('calendar-date');

    fireEvent.click(dateElement);

    const popoverElement = getByTestId('calendar-popover');
    expect(popoverElement).toBeInTheDocument();
  });

  it('should match with snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });
});
