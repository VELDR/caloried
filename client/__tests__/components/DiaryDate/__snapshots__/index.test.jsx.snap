// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DiaryDate Component should match with snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body
    style=""
  >
    <div>
      <div
        class="container"
        data-testid="diary-date"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="ChevronLeftIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M15.41 7.41 14 6l-6 6 6 6 1.41-1.41L10.83 12z"
          />
        </svg>
        <div
          aria-label="Pilih Tanggal"
          class="date"
          data-mui-internal-clone-element="true"
          data-testid="calendar-date"
        >
          Sunday, January 1, 2023
        </div>
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="ChevronRightIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M10 6 8.59 7.41 13.17 12l-4.58 4.59L10 18l6-6z"
          />
        </svg>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="container"
      data-testid="diary-date"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
        data-testid="ChevronLeftIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M15.41 7.41 14 6l-6 6 6 6 1.41-1.41L10.83 12z"
        />
      </svg>
      <div
        aria-label="Pilih Tanggal"
        class="date"
        data-mui-internal-clone-element="true"
        data-testid="calendar-date"
      >
        Sunday, January 1, 2023
      </div>
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
        data-testid="ChevronRightIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M10 6 8.59 7.41 13.17 12l-4.58 4.59L10 18l6-6z"
        />
      </svg>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
